### PR TITLE
[V1][ModelRunner] Support pooling model for v1 engine

### PR DESCRIPTION
**Pull request by Potabk**
## Background

  vLLM V1 engine on NPU currently only supports generative (causal LM) tasks. Pooling models (e.g., Qwen3-Embedding-0.6B for embedding/retrieval) require a different inference path: instead of sampling next tokens, the model's hidden states are pooled into fixed-size vectors. This PR adds pooling task support to the V1 engine with minimal changes to existing code.

  ## Design Overview

  ### 1. npu_input_batch — Decoupled InputBatch for NPU

  **Why fork upstream `gpu_input_batch.py`?**
  The upstream `vllm.v1.worker.gpu_input_batch` interface changes frequently across vLLM versions. To decouple from upstream churn and enable NPU-specific extensions (pooling support), we fork it as `vllm_ascend/worker/npu_input_batch.py`.

  **Key data structures:**

  - `CachedRequestState`: Extended with an optional `pooling_params` field alongside the existing `sampling_params`. For pooling requests, `sampling_params` is None and `pooling_params` is set; for generation requests, the reverse holds. The dataclass also tracks prompt/output token IDs, block IDs, computed token count, and optional LoRA/M-RoPE metadata.

  - `InputBatch`: The core batch state manager. Key additions for pooling:
    - `pooling_params: dict[str, PoolingParams]` — maps req_id to its PoolingParams for all active pooling requests in the batch.
    - `pooling_metadata` property — constructs a `PoolingMetadata` object on-the-fly from the current batch state (prompt lengths, prompt token IDs, and per-request pooling params), used by the model's pooler head.
    - `add_request()` — branching logic: if the request has `sampling_params`, it populates sampling-related buffers (temperature, top_p/k, penalties, etc.); otherwise it asserts `pooling_params` is present and stores it in the `pooling_params` dict.
    - `remove_request()` / `condense()` — properly clean up pooling_params entries when requests complete or batch indices are compacted.

    The remaining InputBatch functionality (token_ids_cpu, num_tokens, block_table, sampling metadata, LoRA mapping, logprobs tracking, etc.) is carried over from upstream gpu_input_batch with NPU-specific adaptations (e.g., no GPU pinned memory assumptions).

  ### 2. PoolingMetadata

  New dataclass at `vllm_ascend/pool/metadata.py`:
  - `prompt_lens: torch.Tensor` — length of each prompt in the batch
  - `prompt_token_ids: Optional[torch.Tensor]` — token IDs (needed by step-pooler models)
  - `pooling_params: list[PoolingParams]` — per-request pooling config

  This mirrors the upstream `vllm.v1.pool.metadata.PoolingMetadata` but is maintained locally for version compatibility.

  ### 3. Model Runner Changes (`model_runner_v1.py`)

  - **Import switch**: `gpu_input_batch` → `npu_input_batch` for `CachedRequestState` and `InputBatch`.
  - **`is_pooling_model`** flag: set when `pooler_config` is present.
  - **`_update_states()`**: Uses `getattr(new_req_data, "pooling_params", None)` for backward compatibility with vLLM v0.9.1 which lacks the field. Sampling params None-check added since pooling requests have no sampling params.
  - **`_process_reqs()`**: Now also returns `num_scheduled_tokens` (np.ndarray) needed by the pooling path to split hidden states per request.
  - **`_pool()`** (new): The pooling inference path:
    1. Splits hidden states by per-request scheduled token counts
    2. Constructs `PoolingMetadata` from InputBatch
    3. Calls `model.pooler()` to produce embeddings
    4. Only emits output for fully-prefilled requests (seq_len == prompt_len)
    5. Returns `ModelRunnerOutput` with `pooler_output` instead of sampled tokens
  - **`execute_model()`**: After `_process_reqs`, checks `input_batch.pooling_params` — if non-empty, routes to `_pool()` instead of the sampling/logits path.
  - **`profile_run()`**: Adds `_dummy_pooler_run()` for memory profiling of pooling models during warmup, constructing dummy PoolingMetadata and invoking the pooler head.
  - **`load_model()`**: Checks for `has_step_pooler` interface (with version compatibility try/except) to set `logits_processing_needs_token_ids`.

  ### 4. Tests

  - **Unit test** (`tests/ut/worker/test_input_batch.py`): Validates InputBatch shapes, add/remove/condense operations, boundary conditions.
  - **E2E test** (`tests/e2e/singlecard/test_embedding.py`): Runs Qwen3-Embedding-0.6B through vLLM and HuggingFace SentenceTransformer, asserts cosine similarity ≥ 0.99 between outputs.
  - **Test utilities**: `HfRunner` class, `check_embeddings_close()`, `matryoshka_fy()` for embedding test infrastructure.

### CODE REVIEW

**Comment by Potabk** [comment]
@Yikun @wangxiyuan I think this should be ready for review

File: vllm_ascend/worker/npu_input_batch.py [hunk 1]
```diff
@@ -0,0 +1,712 @@
+from dataclasses import dataclass
+from typing import Optional, cast
+
+import numpy as np
+import torch
+from vllm.lora.request import LoRARequest
+from vllm.multimodal.inputs import MultiModalKwargs, PlaceholderRange
+from vllm.pooling_params import PoolingParams
+from vllm.sampling_params import SamplingParams, SamplingType
+from vllm.utils import swap_dict_values
```
Comment (wangxiyuan) [inline vllm_ascend/worker/npu_input_batch.py]: conbine with L14

File: vllm_ascend/worker/model_runner_v1.py [hunk 2]
```diff
@@ -80,6 +80,14 @@
 from vllm_ascend.worker.eagle_proposer_v1 import EagleProposer
 from vllm_ascend.worker.mtp_proposer_v1 import MtpProposer
 
+if vllm_version_is("0.9.1"):
```
Comment (wangxiyuan) [inline vllm_ascend/worker/model_runner_v1.py]: is this necessary? `vllm_ascend.worker.npu_input_batch` can't support 0.9.1 vllm?

**Review by wangxiyuan — COMMENTED** [review]
Please add the ut for npu_input_batch as well

File: vllm_ascend/worker/model_runner_v1.py (hunk 2 above)
Comment (Potabk) [reply vllm_ascend/worker/model_runner_v1.py]: `input_batch` in the vllm v0.9.1 has been verified and  will not make any broken changes, if we use vllm_ascend customized interface, then we also need some version compatibility for vllm v0.9.1

Comment (Potabk) [reply vllm_ascend/worker/model_runner_v1.py]: such as ` from vllm.v1.pool.metadata import PoolingMetadata` in the `input_batch`, it is needed to make version compatibility

**Review by wangxiyuan — APPROVED** [review]

**Comment by wangxiyuan** [comment]
ready for review for long time. I'll merge this first
